gamma/utils.py

```diff
@@ -90,62 +90,90 @@ def association(picks, stations, config, event_idx0=0, method="BGMM", **kwargs):
             timestamp0,
             vel,
             method,
-            events,
-            assignment,
             event_idx,
         )
     else:
         manager = mp.Manager()
-        events = manager.list([])
-        assignment = manager.list([])  ## from picks to events
         lock = manager.Lock()
-        event_idx = manager.Value("i", event_idx0)
+        # event_idx0 - 1 as event_idx is increased before use
+        event_idx = manager.Value("i", event_idx0 - 1)
 
         print(f"Associating {len(unique_labels)} clusters with {config['ncpu']} CPUs")
-        with mp.Pool(config["ncpu"]) as p:
-            p.starmap(
+
+        # the following sort and shuffle is to make sure jobs are distributed evenly
+        counter=Counter(labels)
+        unique_labels = sorted(unique_labels, key=lambda x: counter[x], reverse=True)
+        # print top 20 labels' sizes with essencial information
+        # print("top 20 labels' sizes:")
+        # print(
+        #     pd.DataFrame(
+        #         {
+        #             "label": unique_labels[:20],
+        #             "size": [counter[x] for x in unique_labels[:20]],
+        #         }
+        #     )
+        # )
+        np.random.shuffle(unique_labels)
+
+        # the default chunk_size is len(unique_labels)//(config["ncpu"]*4), which makes some jobs very heavy
+        chunk_size = max(len(unique_labels)//(config["ncpu"]*20), 1)
+        with mp.get_context('spawn').Pool(config["ncpu"]) as p:
+            results = p.starmap(
                 associate,
                 [
                     [
-                        data[labels == k],
```

Comment:
What is the main difference between passing [k, labels, data, locs...] v.s. [data[labels ==k], locs[labels==k]...]?

Reply:
When processing a large dataset, it takes more than 1 hour to finish the loop. It can save some time passing the k.